tools/extract_packets.py

```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
+
+"""
+This tool extracts all packets of a specified protocol layer from a PCAP file and writes the byte
+representation of each packet into a separate file.
```

Comment:
The tool in its current form has not been very useful for me. I tried to extract a couple of protocols I needed recently:

- PPPoE (seems to be unsupported by scapy)
- HTTP/2 (seems to be unsupported by scapy)
- IPv6 (I used the "inet6" layer, but nothing was dumped)

I have the feeling that we won't get very far with Scapy. Probably something based on Wireshark would have better chances of supporting protocols we are interested in.

In general I wonder whether it would be better to extract the payload of an upper layer protocol. This would allow us to request the payload field of Ethernet to be extracted whenever Ether_Type is 0x86dd when we want IPv6 packets.

Reply:
>     * IPv6 (I used the "inet6" layer, but nothing was dumped)

A specific class in a Scapy layer package has to be given:

```
usage: extract_packets.py [-h] LAYER PCAP_FILE OUTPUT_DIRECTORY

positional arguments:
  LAYER             Scapy layer in the form <package>.<class> (e.g., dhcp.BOOTP)
  PCAP_FILE
  OUTPUT_DIRECTORY
```

For IPv6 the layer argument must be `inet6.IPv6`.

> I have the feeling that we won't get very far with Scapy. Probably something based on Wireshark would have better chances of supporting protocols we are interested in.

Creating a tool based on Scapy was a very easy and fast solution for me and it worked very well for DHCP. There is still a number of [protocols](https://scapy.readthedocs.io/en/latest/api/scapy.layers.html) for which it could be useful for us in the future. Nevertheless, you are probably right and looking into a more powerful alternative would make sense.

> In general I wonder whether it would be better to extract the payload of an upper layer protocol. This would allow us to request the payload field of Ethernet to be extracted whenever Ether_Type is 0x86dd when we want IPv6 packets.

I think the current approach works fine for all layers directly supported by Scapy, but maybe your described approach could be helpful to extract unsupported layers as well.

Reply:
I have improved the argument validation in c089c86. `inet6` is detected as an invalid layer argument now.

Reply:
I have added an option to show a list of all available Scapy layers in 1fed3a3.

Reply:
That option is very helpful.